### ISSUE TITLE
Freezing `ssh` connections.

### ISSUE THREAD

**Issue by drawkula** [minoca/os#62]
Running Minoca OS x86dbg in `qemu` often shows stuck or frozen `ssh` connections.
```
(yeti@kumari:11)/wrk/minoca/qemu$ ./run -drive file=system.32G.qcow2
(yeti@kumari:11)/wrk/minoca/qemu$ ssh -p22222 root@localhost
root@localhost's password: 
~# uname -a
Minoca minoca 0.3.0.1964-beta-debug broot-c9a6467 Nov 23 2016 10:58:13 i686
```
This is https://github.com/minoca/os/commit/4e4eea96be5c2e0a23c7f571398a735c1a2248f0 plus "GMT patch".
```
~# zcat /src/third-party/src/gcc-4.9.2.tar.gz | tar tv
```
...reproducibly freezes the `ssh` session.

----

`run` is my `qemu` launch helper:
```
#!/bin/sh
exec qemu-system-x86_64 \
        -enable-kvm \
        -smp 2 \
        -m 1024 \
        -net nic,model=i82559er,macaddr=fe:fd:0a:00:02:0e \
        -net user,hostfwd=tcp::22222-:22 \
        "$@" \
        &
```

**Comment by ccstevens** [comment]
Thanks for the repro steps. I couldn't hit it on the RPI 3, but reproduced it on QEMU. It looks like a deadlock. `tar` is writing to the slave terminal, but filled up the internal buffer space and started waiting for the master terminal to read. `sshd` has the handle to the master terminal but is stuck polling on a pipe and a socket. Not yet sure why this poll isn't getting satisfied. I'd expect network traffic to come in.

**Comment by ccstevens** [comment]
@drawkula Sorry for the delay on this. Finally tracked it down. It ended up being a bug in QEMU's emulation of the e100 network card. If QEMU ever transmits 16 packets in a row, it stops its transmit engine (a while-loop), but does not set the state to "suspended" or interrupt to tell the system it's no longer active. So, the Minoca e100 driver was seeing the transmit engine's state as "active" and thought it didn't need to resume the transmit engine when sending more packets. As a result, the send side of networking came to a complete stand-still. I worked around the QEMU issue with a driver fix (forcing a resume with each send). 

The fix will be pushed to github in the next batch of changes - sometime this week.

**Comment by drawkula** [comment]
Thanks for catching this beast!

Reading your answer I thought switching to `-nic model=rtl8139` would be a fix if it is only an `e100` problem but it (or something very similar from the user's perspective) seems to be there for `rtl8139` too... :.(

Is the problem in parts of Qemu's code that's shared between different NICs?

**Comment by ccstevens** [comment]
Hmm. The QEMU code was specific to the `e100` NIC. I didn't try `rtl8139`, but I did try `pcnet` and didn't encounter any issues. I guess I'll give `rtl8139` a try.

**Comment by drawkula** [comment]
Sorry... mixed my startscripts... I got the last hangs with `model=pcnet`.... I'll update the original message...

**Comment by drawkula** [comment]
Have hit the wrong button... damn flu!
Reopening...

**Comment by ccstevens** [comment]
Didn't realize we'd pushed updates to github, but the `i82559er`/`e100` fix is in change [838c2965](https://github.com/minoca/os/commit/838c296569c8e8bb523ff4bd71a4c30db9068c01).

**Comment by drawkula** [comment]
I'm running https://github.com/minoca/os/commit/d7bc7915a90abfb7e8555a01ff9bf512acda7856 plus TZ changed to GMT and modified console colours which both should be harmless.

**Comment by ccstevens** [comment]
@drawkula, try [6243af9d](https://github.com/minoca/os/commit/6243af9d0656d9e4d2d2ef9c633bf283b85e9303). This got rid of the `i82559er` hang I was seeing with my `echo` script.

Still need to look at `pcnet`. Would be curious if you ever see that again.

**Comment by drawkula** [comment]
Using https://github.com/minoca/os/commit/6243af9d0656d9e4d2d2ef9c633bf283b85e9303 (plus GMT and terminal color changes) `i82559er` survived 100 "tar-tests".

**Comment by ccstevens** [comment]
@drawkula You weren't crazy when you saw the `pcnet` driver hanging. I was able to reproduce it and track down the issue. Reading/writing to `pcnet` registers requires programming an address port and then reading/writing a data port. These two steps were not synchronized/atomic. My guess is that the driver was corrupting register state, causing the hang. The change will be in the next batch pushed to GitHub, but don't feel compelled to verify it. It's a tough hang to hit and I'm not seeing it anymore after 30 mins of SSH spew.

Closing this one out.
